File: frontend/browser/store/rice.ts
// TODO:
// [ ] activeProject, cookStatus, cookErrors, modelStatus — https://docs.pmnd.rs/zustand
// [ ] startCook NATS cook.reload.{project}; stopCook; setModelStatus
//
import { create } from "zustand";

type RiceRuntimeState = {
	values: Record<string, unknown>;
	set: (key: string, value: unknown) => void;
	get: (key: string) => unknown;
	reset: () => void;
	hydrate: (data: Record<string, unknown>) => void;
};

export const useRiceStore = create<RiceRuntimeState>((set, get) => ({
	values: {},
	set: (key, value) =>
		set((s) => ({
			values: { ...s.values, [key]: value },
		})),
	get: (key) => get().values[key],
	reset: () => set({ values: {} }),
	hydrate: (data) =>
		set((s) => ({
			values: { ...s.values, ...data },
		})),
}));

File: frontend/screen/lib/store/rice.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

// TODO:
// [ ] activeProject, cookStatus, cookErrors; Phoenix invalidation
//
class RiceState {
  const RiceState({this.values = const <String, Object?>{}});

  final Map<String, Object?> values;

  RiceState copyWith({Map<String, Object?>? values}) => RiceState(values: values ?? this.values);
}

class RiceNotifier extends StateNotifier<RiceState> {
  RiceNotifier() : super(const RiceState());

  Object? get(String key) => state.values[key];

  void set(String key, Object? value) {
    state = state.copyWith(values: {...state.values, key: value});
  }

  void hydrate(Map<String, Object?> data) {
    state = state.copyWith(values: {...state.values, ...data});
  }

  void reset() {
    state = const RiceState();
  }
}
